File: app/src/main/java/com/textfield/json/ottawastreetcameras/MainActivity.java
package com.textfield.json.ottawastreetcameras;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.SearchView;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.util.Log;
import android.view.Gravity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AbsListView;
import android.widget.AdapterView;
import android.widget.LinearLayout;
import android.widget.LinearLayout.LayoutParams;
import android.widget.ListView;
import android.widget.TextView;

import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonArrayRequest;
import com.android.volley.toolbox.Volley;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;

public class MainActivity extends AppCompatActivity {
    ArrayList<Camera> cameras = new ArrayList<>();
    File cameraJson;
    CameraAdapter myAdapter;
    ListView cameraListView;
    private ArrayList<Camera> selectedCameras = new ArrayList<>();

    public void downloadJson() {

        String url = "http://traffic.ottawa.ca/map/camera_list";
        RequestQueue queue = Volley.newRequestQueue(this);
        JsonArrayRequest jsObjRequest = new JsonArrayRequest(url, new Response.Listener<JSONArray>() {
            @Override
            public void onResponse(JSONArray response) {
                try {
                    FileWriter fileWriter = new FileWriter(cameraJson);
                    fileWriter.write(response.toString());
                    fileWriter.flush();
                    setup();
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        }, new Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError error) {

            }
        });
        queue.add(jsObjRequest);

    }

    public void setup() {
        cameraListView = (ListView) findViewById(R.id.listView);

        final Toolbar toolbar = (Toolbar) findViewById(R.id.main_toolbar);
        toolbar.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                cameraListView.smoothScrollToPosition(0);
            }
        });
        setSupportActionBar(toolbar);

        try {
            JSONArray jsonArray = new JSONArray(loadJSONFromFile());
            for (int i = 0; i < jsonArray.length(); i++) {
                JSONObject g = (JSONObject) jsonArray.get(i);
                cameras.add(new Camera(g));
            }


            Collections.sort(cameras);
            myAdapter = new CameraAdapter(this, cameras);
            cameraListView.setAdapter(myAdapter);

            final ArrayList<String> indexTitles = myAdapter.getIndexTitles();
            final HashMap<String, Integer> index = myAdapter.getIndex();

            LinearLayout linearLayout = (LinearLayout) findViewById(R.id.indexHolder);
            for (int i = 0; i < index.size(); i++) {
                TextView t = new TextView(this);
                LayoutParams layoutParams = new LayoutParams(LayoutParams.MATCH_PARENT, 0, 1);
                layoutParams.gravity = Gravity.CENTER_HORIZONTAL;

                t.setLayoutParams(layoutParams);
                t.setText(indexTitles.get(i));
                t.setTextColor(ContextCompat.getColor(this, R.color.colorAccent));
                t.setGravity(Gravity.CENTER_HORIZONTAL);
                t.setBackgroundColor(ContextCompat.getColor(this, android.R.color.transparent));

                final int finalI = i;
                t.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        cameraListView.setSelection(index.get(indexTitles.get(finalI)));
                    }
                });
                linearLayout.addView(t);
            }


            cameraListView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
                @Override
                public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
                    Bundle b = new Bundle();
                    ArrayList<Camera> cams = new ArrayList<>(Arrays.asList(new Camera[]{myAdapter.getItem(i)}));
                    b.putParcelableArrayList("cameras", cams);

                    Intent intent = new Intent(MainActivity.this, CameraActivity.class);
                    intent.putExtras(b);
                    startActivity(intent);
                }
            });
            cameraListView.setItemsCanFocus(true);

            cameraListView.setMultiChoiceModeListener(new AbsListView.MultiChoiceModeListener() {
                @Override
                public void onItemCheckedStateChanged(android.view.ActionMode actionMode, int i, long l, boolean b) {
                    //cameraListView.getSelectedView().setSelected(b);
                    if (b) {
                        selectedCameras.add(myAdapter.getItem(i));
                    } else {
                        selectedCameras.remove(myAdapter.getItem(i));
                    }
                }

                @Override
                public boolean onCreateActionMode(android.view.ActionMode actionMode, Menu menu) {
                    actionMode.getMenuInflater().inflate(R.menu.contextual_menu, menu);
                    return true;
                }

                @Override
                public boolean onPrepareActionMode(android.view.ActionMode actionMode, Menu menu) {
                    return false;
                }

                @Override
                public boolean onActionItemClicked(android.view.ActionMode actionMode, MenuItem menuItem) {
                    if (menuItem.getItemId() == R.id.open_cameras) {
                        Intent intent = new Intent(MainActivity.this, CameraActivity.class);
                        intent.putParcelableArrayListExtra("cameras", selectedCameras);
                        startActivity(intent);
                        return true;
                    }
                    return false;
                }

                @Override
                public void onDestroyActionMode(android.view.ActionMode actionMode) {
                    selectedCameras.clear();
                }
            });

        } catch (JSONException e) {
            //e.printStackTrace();
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        cameraJson = new File(getCacheDir() + "/cameras.json");
        if (!cameraJson.exists()) {
            Log.w("FILE:", "Does not exist");
            downloadJson();
        } else {
            setup();
        }


        /*DB myDB = new DB(this);
        myDB.createDatabase();
        myDB.open();
        Cursor cursor = myDB.runQuery("select * from cameras;");

        do {
            Camera camera = new Camera(cursor);
            cameras.add(camera);
        } while (cursor.moveToNext());
        myDB.close();

        Collections.sort(cameras);*/



        /*try {
            JSONArray jsonArray = new JSONArray(loadJSONFromFile());
            for (int i = 0; i < jsonArray.length(); i++) {
                JSONObject g = (JSONObject) jsonArray.get(i);
                cameras.add(new Camera(g));
            }


        } catch (JSONException e) {
            //e.printStackTrace();
        }*/


    }

    public String loadJSONFromFile() {
        String json;
        try {
            InputStream is = new FileInputStream(cameraJson);
            int size = is.available();
            byte[] buffer = new byte[size];
            is.read(buffer);
            is.close();
            json = new String(buffer, "UTF-8");
        } catch (IOException ex) {
            //ex.printStackTrace();
            return null;
        }
        return json;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.menuItemMap) {
            Intent intent = new Intent(this, MapsActivity.class);
            intent.putParcelableArrayListExtra("cameras", cameras);
            startActivity(intent);
        }
        return true;
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        SearchView searchView = (SearchView) MenuItemCompat.getActionView(menu.findItem(R.id.user_searchView));
        searchView.setQueryHint(String.format(getResources().getString(R.string.search_hint), cameras.size()));

        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                if (TextUtils.isEmpty(newText)) {
                    myAdapter.getFilter().filter("");
                    cameraListView.clearTextFilter();
                } else {
                    myAdapter.getFilter().filter(newText);
                }
                return true;
            }
        });

        return true;
    }
}

File: app/src/main/java/com/textfield/json/ottawastreetcameras/MainActivity.kt
package com.textfield.json.ottawastreetcameras

import android.content.Intent
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.support.v4.view.MenuItemCompat
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.SearchView
import android.support.v7.widget.Toolbar
import android.text.TextUtils
import android.util.Log
import android.view.Gravity
import android.view.Menu
import android.view.MenuItem
import android.widget.*
import android.widget.LinearLayout.LayoutParams
import com.android.volley.Response
import com.android.volley.toolbox.JsonArrayRequest
import com.android.volley.toolbox.Volley
import kotlinx.android.synthetic.main.activity_main.*
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.File
import java.io.FileInputStream
import java.io.FileWriter
import java.io.IOException
import java.nio.charset.Charset
import java.util.*

class MainActivity : AppCompatActivity() {
    internal var cameras = ArrayList<Camera>()
    internal lateinit var cameraJson: File
    internal lateinit var myAdapter: CameraAdapter
    internal lateinit var cameraListView: ListView
    private val selectedCameras = ArrayList<Camera>()

    fun downloadJson() {

        val url = "http://traffic.ottawa.ca/map/camera_list"
        val queue = Volley.newRequestQueue(this)
        val jsObjRequest = JsonArrayRequest(url, Response.Listener { response ->
            try {
                val fileWriter = FileWriter(cameraJson)
                fileWriter.write(response.toString())
                fileWriter.flush()
                setup()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }, Response.ErrorListener { })
        queue.add(jsObjRequest)

    }

    fun setup() {

        cameraListView = listView

        main_toolbar.setOnClickListener { cameraListView.smoothScrollToPosition(0) }
        setSupportActionBar(main_toolbar)

        try {
            val jsonArray = JSONArray(loadJSONFromFile())
            for (i in 0 until jsonArray.length()) {
                val g = jsonArray.get(i) as JSONObject
                cameras.add(Camera(g))
            }


            Collections.sort(cameras)
            myAdapter = CameraAdapter(this, cameras)
            cameraListView.adapter = myAdapter

            val indexTitles = myAdapter.getIndexTitles()
            val index = myAdapter.getIndex()

            val linearLayout = findViewById(R.id.indexHolder) as LinearLayout?
            for (i in 0 until index.size) {
                val t = TextView(this)
                val layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, 0, 1f)
                layoutParams.gravity = Gravity.CENTER_HORIZONTAL

                t.layoutParams = layoutParams
                t.text = indexTitles[i]
                t.setTextColor(ContextCompat.getColor(this, R.color.colorAccent))
                t.gravity = Gravity.CENTER_HORIZONTAL
                t.setBackgroundColor(ContextCompat.getColor(this, android.R.color.transparent))

                t.setOnClickListener { cameraListView.setSelection(index[indexTitles[i]]!!) }
                linearLayout!!.addView(t)
            }


            cameraListView.onItemClickListener = AdapterView.OnItemClickListener { adapterView, view, i, l ->
                val b = Bundle()
                val cams = ArrayList(Arrays.asList(myAdapter.getItem(i)!!))
                b.putParcelableArrayList("cameras", cams)

                val intent = Intent(this@MainActivity, CameraActivity::class.java)
                intent.putExtras(b)
                startActivity(intent)
            }
            cameraListView.itemsCanFocus = true

            cameraListView.setMultiChoiceModeListener(object : AbsListView.MultiChoiceModeListener {
                override fun onItemCheckedStateChanged(actionMode: android.view.ActionMode, i: Int, l: Long, b: Boolean) {
                    //cameraListView.getSelectedView().setSelected(b);
                    if (b) {
                        selectedCameras.add(myAdapter.getItem(i)!!)
                    } else {
                        selectedCameras.remove(myAdapter.getItem(i))
                    }
                }

                override fun onCreateActionMode(actionMode: android.view.ActionMode, menu: Menu): Boolean {
                    actionMode.menuInflater.inflate(R.menu.contextual_menu, menu)
                    return true
                }

                override fun onPrepareActionMode(actionMode: android.view.ActionMode, menu: Menu): Boolean {
                    return false
                }

                override fun onActionItemClicked(actionMode: android.view.ActionMode, menuItem: MenuItem): Boolean {
                    if (menuItem.itemId == R.id.open_cameras) {
                        val intent = Intent(this@MainActivity, CameraActivity::class.java)
                        intent.putParcelableArrayListExtra("cameras", selectedCameras)
                        startActivity(intent)
                        return true
                    }
                    return false
                }

                override fun onDestroyActionMode(actionMode: android.view.ActionMode) {
                    selectedCameras.clear()
                }
            })

        } catch (e: JSONException) {
            //e.printStackTrace();
        }

    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        cameraJson = File(cacheDir.toString() + "/cameras.json")
        if (!cameraJson.exists()) {
            Log.w("FILE:", "Does not exist")
            downloadJson()
        } else {
            setup()
        }


        /*DB myDB = new DB(this);
        myDB.createDatabase();
        myDB.open();
        Cursor cursor = myDB.runQuery("select * from cameras;");

        do {
            Camera camera = new Camera(cursor);
            cameras.add(camera);
        } while (cursor.moveToNext());
        myDB.close();

        Collections.sort(cameras);*/


        /*try {
            JSONArray jsonArray = new JSONArray(loadJSONFromFile());
            for (int i = 0; i < jsonArray.length(); i++) {
                JSONObject g = (JSONObject) jsonArray.get(i);
                cameras.add(new Camera(g));
            }


        } catch (JSONException e) {
            //e.printStackTrace();
        }*/


    }

    fun loadJSONFromFile(): String? {
        val json: String
        try {
            val `is` = FileInputStream(cameraJson)
            val size = `is`.available()
            val buffer = ByteArray(size)
            `is`.read(buffer)
            `is`.close()
            json = String(buffer, Charset.forName("UTF-8"))
        } catch (ex: IOException) {
            //ex.printStackTrace();
            return null
        }

        return json
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.menuItemMap) {
            val intent = Intent(this, MapsActivity::class.java)
            intent.putParcelableArrayListExtra("cameras", cameras)
            startActivity(intent)
        }
        return true
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        val searchView = MenuItemCompat.getActionView(menu.findItem(R.id.user_searchView)) as SearchView
        searchView.queryHint = String.format(resources.getString(R.string.search_hint), cameras.size)

        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {
                if (TextUtils.isEmpty(newText)) {
                    myAdapter.filter.filter("")
                    cameraListView.clearTextFilter()
                } else {
                    myAdapter.filter.filter(newText)
                }
                return true
            }
        })

        return true
    }
}
